### DISCUSSION TITLE
multi-domain site with i18n

### DISCUSSION THREAD

**Discussion by bonesoul** [vercel/next.js#38607] [Help]
so i've followed the guides over https://nextjs.org/docs/advanced-features/i18n-routing and https://nextjs.org/docs/advanced-features/middleware to create a multi domain site with different/standalone languages on their own.

basically my config is:

```
  i18n: {
    locales: ['en', 'tr'],
    defaultLocale: 'en',
    localeDetection: false, // disable automatic locate detection.
    domains: [
      {
        domain: 'englishsite.com',
        defaultLocale: 'en',
      },
      {
        domain: 'turkishsite.com',
        defaultLocale: 'tr',
      }
    ]
  },
```

what i'd like to have is a two seperate sites from a single next.js codebase:

- englishsite.com with 'en' locale and only english routes / pages.
- turkishsite.com with 'tr' locate and only turkish routes /pages.

so basically:
note: 'hakkinda' is the turkish word for about.

- englishsite.com/about [✔️ - valid route]
- englishsite.com/hakkinda [❌- invalid route]
- turkishsite.com/about [❌ - invalid route]
- turkishsite.com/hakkinda  [✔️ - valid route]

And i basically want all of valid routes to be processed by pages/about.tsx page, probably with a rewrite?

So which route should i take? 
- rewrites with i18n support? (https://nextjs.org/docs/api-reference/next.config.js/rewrites#rewrites-with-i18n-support)
- or a middleware approach?

I tried to first once with no luck but have some progress with the middleware approach:

```
import { NextRequest, NextResponse } from "next/server";
const public_file_regex = /\.(.*)$/;

export async function middleware(req: NextRequest) {
  // if serving a publif file or api path, just skip the rest.
  if (
    req.nextUrl.pathname.startsWith("/_next") ||
    req.nextUrl.pathname.includes("/api/") ||
    public_file_regex.test(req.nextUrl.pathname)
  )
    return;

  if (
    req.nextUrl.locale === "tr" &&
    req.nextUrl.pathname.startsWith("/hakkinda")
  )
    return NextResponse.rewrite(new URL("/about", req.url));
}

// urls to run the middleware against.
export const config = {
  matcher: ["/hakkinda"],
};

```

Any ideas or guides to achieve this?

**Comment by federicovezzoli** [comment]
I'm trying to do the exact thing, and what I accomplished so far is that even with this configuration: 
```  i18n: {
    locales: ['en-gb', 'es-es', 'de-de', 'fr-fr', 'it-it'],
    defaultLocale: 'it-it',
    domains: [
    {
      domain: 'xxx-it.netlify.app',
      defaultLocale: 'it-it'
    },
    {
      domain: 'xxx-en.netlify.app',
      defaultLocale: 'en-gb'
    },
    {
      domain: 'xxx-fr.netlify.app',
      defaultLocale: 'fr-fr'
    },
    {
      domain: 'xxx-de.netlify.app',
      defaultLocale: 'de-de'
    },
    {
      domain: 'xxx-es.netlify.app',
      defaultLocale: 'es-es'
    }
],
    localeDetection: false
  },
  ```
  
  it just generates all the routes with the standard sub-path routing:
  `xxx/en-gb/slug
  xxx/it-it/slug`
  
  Anyone can help? is this a bug? I basically couldn't find anything about that.

**Comment by bonesoul** [comment]
I eventually dropped the locale support and just use this middleware to rewrite routes for localized versions of my english site.

```
middleware.ts

import { Language, SiteLanguage } from "lib/i18n";
import { NextRequest, NextResponse } from "next/server";
const PUBLIC_FILE_REGEX = /\.(.*)$/;

const routes: LanguageRouteRewrite[] = [
  {
    source: "/hakkinda",
    destination: "/about",
    locale: "tr",
  },
];

export async function middleware(req: NextRequest) {
  if (
    req.nextUrl.pathname.startsWith("/_next") || // build folder.
    req.nextUrl.pathname.includes("/api/") || // apis.
    PUBLIC_FILE_REGEX.test(req.nextUrl.pathname) || // public path.
    SiteLanguage == Language.en // or we are in international mode.
  )
    return; // skip any rewrites for i18n handling.

  for (const entry of routes) {
    // check if we need a rewrite from source to destination.
    if (req.nextUrl.pathname.startsWith(entry.source))
      return NextResponse.rewrite(new URL(entry.destination, req.url));

    // make sure we english routes are not usable in non-english versions.
    if (req.nextUrl.pathname.startsWith(entry.destination))
      return NextResponse.rewrite(new URL("/404", req.url));
  }
}

// urls to run the middleware against.
export const config = {
  matcher: ["/about", "/hakkinda"],
};

interface LanguageRouteRewrite {
  source: string;
  destination: string;
  locale: string;
}
```

and 

```
lib/i18n.ts

/**
 * available site languages.
 */
export enum Language {
  en = "en",
  tr = "tr",
}

/**
 * returns site language.
 */
export const SiteLanguage: Language = process.env.NEXT_PUBLIC_SITE_LANGUAGE === "tr" ? Language.tr : Language.en;

/**
 * returns if site is in international mode.
 */
// @ts-ignore: allow overlap.
export const IsEn = SiteLanguage === Language.en;
```

**Reply by federicovezzoli** [reply]
yeah I thought of that as well, but don't have any experience with middlewares so I thought to stick to the standard implementation...

how does your config file look like? in this configuration?
How do you handle the links to the different version of each page?

**Reply by bonesoul** [reply]
```
const routes: LanguageRouteRewrite[] = [
  {
    source: "/hakkinda",
    destination: "/about",
    locale: "tr",
  },
];
```

about.tsx is the default english version of the page. when the middleware is aware of a hit to /hakkinda (turkish version), it rewrites the request to about.tsx.

```
    // check if we need a rewrite from source to destination.
    if (req.nextUrl.pathname.startsWith(entry.source))
      return NextResponse.rewrite(new URL(entry.destination, req.url));

    // make sure we english routes are not usable in non-english versions.
    if (req.nextUrl.pathname.startsWith(entry.destination))
      return NextResponse.rewrite(new URL("/404", req.url));
```

with the second line it make sures /about route is also not usable in turkish version.

my next.config.js is very standard one that doesn't utilize locales.

**Reply by federicovezzoli** [reply]
how do you handle the generation of the links for an eventual language switcher?

**Reply by Derick1530** [reply]
Sir. do you mind sharing the next.config? I working on a project I need to find a hack to make what you are trying to do here
